### ISSUE TITLE
Permit non-void return types for partial methods

### ISSUE THREAD

**Issue by AaronRobinsonMSFT** [dotnet/csharplang#3301]
The Source Generator feature is being investigated to create a runtime agnostic and AOT compatible P/Invoke stub generator. One of the hard requirements of Source Generators is user written code is read only and cannot be altered by any Source Generator implementation. This requirement makes it difficult to declare methods and use them, but have a Source Generator provide an implementation in the future (i.e. build time).

Possibly work-arounds exist by declaring a partial class and having a convention.

``` CSharp
// User defined
[FutureMethod("Foo", typeof(int))]
partial class GenClasses
{
}

// Generated in another TU
partial class GenClasses
{
   public static int Foo() { ... }
}
```

The above could be made to work, but there are issues.

* How does documentation for `Foo` work?
* What is the IDE scenario here for IntelliSense?
* Users would probably see many red squiggles that make development annoying.

An alternative approach would be a type of forward declaration using [`partial`](https://docs.microsoft.com/dotnet/csharp/language-reference/keywords/partial-method) methods. The below would relax the requirement to have a `void` return type and thus the generation of the corresponding method could occur. IntelliSense would make sense as would a [location for XML documentation](https://github.com/dotnet/runtime/pull/33742#issuecomment-602290467) in the user defined code. Note that `partial` methods are also required to be `private`. The below example does adhere to that requirement. Relaxing that would also be beneficial but not strictly needed.

``` CSharp
// User defined
partial class GenClasses
{
   static partial int FooImpl();
   public static Foo() { return FooImpl(); }
}

// Generated in another TU
partial class GenClasses
{
   static partial int FooImpl() { ... }
}
```

/cc @jaredpar @davidwrighton @jkotas @stephentoub

**Comment by YairHalberstadt** [comment]
I presume that unlike void private methods, an implementation of the partial method would be required?

**Comment by AaronRobinsonMSFT** [comment]
@YairHalberstadt I personally have no objections to that, but it would be up to the language teams to make the final decision. I would imagine some complexity if that wasn't the case though.

**Comment by agocke** [comment]
I independently thought of the same thing when @chsienki and I were brainstorming a while ago. Seems useful. I assumed that non-void return types would require an implementation method.

**Comment by jaredpar** [comment]
> I presume that unlike void private methods, an implementation of the partial method would be required?

Yes indeed. It would also mean that `partial` methods which have non-void return methods cannot be erased at the call site; there will always be an invocation of the method. 

> partial methods are also required to be private. The below example does adhere to that requirement, but relaxing that would also be beneficial although not strictly needed.

I think we'd also want to lift that restriction. Think the best way to approach that part is to say that `partial` methods which have a non-private accessibility are required to have an implementation. That gives it similar rules as `non-void` return types.

**Comment by tannergooding** [comment]
If `private void` partial-methods are not required, and `non-void` partial methods (any accessibility) are required, would `non-private void` partial methods be required?

Would it be worth something allowing the user to specify that the `void` partial method (any accessibility) is required to help catch issues where they expect it to be provided by a source generator and want to error if it isn't?

**Comment by 333fred** [comment]
`non-private void` should almost certainly be required. They're part of a public api, and having them silently do nothing would break abstraction boundaries.

**Comment by jaredpar** [comment]
@tannergooding 

> If private void partial-methods are not required, and non-void partial methods (any accessibility) are required, would non-private void partial methods be required?

Yes. As my comment said: partial methods which have a non-private accessibility are required to have an implementation. I deliberately left out return type in that statement. 

> Would it be worth something allowing the user to specify that the void partial method (any accessibility) is required to help catch issues where they expect it to be provided by a source generator and want to error if it isn't?

Possibly but I'd wait to see a compelling case before adding a new feature to support this. So far we've been discussing relaxing existing restrictions to support source generators. This would be adding a new restriction and is essentially a new feature (need syntax or attribute to support).

**Comment by alrz** [comment]
Previous discussion: https://github.com/dotnet/csharplang/issues/753

**Comment by AaronRobinsonMSFT** [comment]
>Previous discussion: #753

@alrz That is definitely being discussed here. Should we close that issue in lieu of this one since this is a super-set?

**Comment by alrz** [comment]
>Should we close that issue in lieu of this one since this is a super-set?

Sure, just wanted to add the link just in case. Closed.

**Comment by AaronRobinsonMSFT** [comment]
@agocke @jaredpar @chsienki Can you please take another look and ensure this plan doesn't contain any other assumptions that could cause issues with using Source Generators?

**Comment by CyrusNajmabadi** [comment]
I don't actually understand the need here for this.  Addressing hte OP:

> How does documentation for Foo work?

The generated Foo includes its documentation.  That documentation flows along the same way as nodemal code.

> What is the IDE scenario here for IntelliSense?

IDE works because it sees the generated code.

> Users would probably see many red squiggles that make development annoying.

That should not happen.  The source generator will run and will have generated the method the user can call.




a

**Comment by AaronRobinsonMSFT** [comment]
> > How does documentation for Foo work?
> 
> The generated Foo includes its documentation. That documentation flows along the same way as nodemal code.

The auto-generated code isn't what the user wrote, so how would it know the documentation for the API? In the below code, where does the XML documentation for the function `Foo` reside? Any existing documentation tool would be confused by documenting the attribute with the typical XML documentation for a method.

```
// User defined
[FutureMethod("Foo", typeof(int))]
partial class GenClasses
{
}
```
 
> > What is the IDE scenario here for IntelliSense?
> 
> IDE works because it sees the generated code.

That is only true if source generators are continually running in the background. It is entirely possible for this to work similar to XAML, but there are terrible cases there especially during a copy/paste of large swathes of XAML where the editor needs to wait for the language service to keep up. There would be squiggles and general confusion since `Foo`'s signature is being generated and defined by the user in a dynamic manner. As opposed to XAML where there is an inheritance contract to conform to and possibly optimize for the limited set of API signatures that would exist.

In the proposed case, there is a function declaration a head of time which satisfies the location for XML documentation and completely addresses the IntelliSense issue without putting undo pressure on the IDE to try and generate potentially large amounts of code during the edit cycle.

**Comment by stephentoub** [comment]
> The generated Foo includes its documentation. That documentation flows along the same way as nodemal code.

Flows from where?

Let's say this is a regex generator.  I want to be able to write this:
```C#
/// <summary>Creates a regex that finds hello followed by world, with anything in the middle</summary>
[RegexGenerator("hello.*world", RegexOptions.IgnoreCase)]
internal partial Regex HelloWorldRegex();
```

With that I don't have to have a string-based name.  I've written the method I'll be calling like any other, except the generator is filling in the implementation, rather than the method magically appearing out of thing air.  I've been able to put documentation on the method like I would on any other method.  The regular expression that's tied to this method is right there in the method's attribute, not somewhere else in the assembly attributing some less specific thing.  Etc.

How does this look without the non-void returning support?

**Comment by CyrusNajmabadi** [comment]
> That is only true if source generators are continually running in the background. 

Yes.  That's the plan here.

**Comment by CyrusNajmabadi** [comment]
> et's say this is a regex generator. I want to be able to write this:

I would not write it like that.  i would just have:

```c#
[RegexGenerator("hello.*world", RegexOptions.IgnoreCase, "HelloWorldRegex")]
partial class Whatever
{
}
```

SourceGenerator would take that and make another part for `Whatever` that has the `HelloWorldRegex` property with teh right code injected for it.  All features looking at Whatever, will see that member, because that member exists.

**Comment by stephentoub** [comment]
Yeah, to me that's much worse, for all the reasons I outlined.

**Comment by CyrusNajmabadi** [comment]
> The auto-generated code isn't what the user wrote, so how would it know the documentation for the API? In the below code,

I'm not really understanding what you mean.  When the generator generates the new code it can put whatever comments on it using whatever scheme you want to use to convey information to your generator.  It could be embedded in attributes, or comments, or something else.

**Comment by CyrusNajmabadi** [comment]
>With that I don't have to have a string-based name. I've written the method I'll be calling like any other, except the generator is filling in the implementation

This sounds like the `original/replace` approach that we decided to cancel.  these scenarios have a suitable approach with the 'generate a new file' system.  It may not be really glamorous, but i don't expect or want it to be.  I'd rather have a simple system that can do this, instead of more and more complex features (like actually changing the language) in order to make this happen.

**Comment by CyrusNajmabadi** [comment]
> With that I don't have to have a string-based name. 

you don't need a string base dname, the following works fine:

`[RegexGenerator("hello.*world", RegexOptions.IgnoreCase, nameof(HelloWorldRegex)]`

there's a duality here.  teh generator can see the above and use it to make the method called `HelloWorldRegex`.  The attribute can also then see that new method and then compile.

**Comment by CyrusNajmabadi** [comment]
> And, to the documentation point, it lacks it.

```
// This is the documentation to generate on the member.  The source generator can look for this
// and move it over.  The source generator can totally make an informal mechanism to allow its domain
// to pass it the information it needs
[RegexGenerator("hello.*world", RegexOptions.IgnoreCase, nameof(HelloWorldRegex)]
```

**Comment by stephentoub** [comment]
> The source generator can totally make an informal mechanism

Why would I want that when we already have a long-standing, compiler-supported mechanism for this?  I don't want to invent my own mechanisms, worry that an arbitrary comment in just the wrong place is going to be misconstrued by the ad-hoc conventions of the generator.

**Comment by CyrusNajmabadi** [comment]
> There would be squiggles and general confusion since Foo's signature is being generated and defined by the user in a dynamic manner. 

That would be weird.  We're not going to run diagnostics on hte code *prior* to source generation.  Expectation is we run it *after* source generation.  IN other words, it's part of our design that the code prior to gen is *normally* expected to be in error because it will depend on the generated code.

In other words, the thing you're trying to fix/avoid here is precisely the use model we're trying heavily to move toward.  the expectation is absolutely to not have the definitions or aything else informing the original source code.  but all IDE features and whatnot only operate on the post-transformed code.

**Comment by CyrusNajmabadi** [comment]
> Why would I want that when we already have a long-standing, compiler-supported mechanism for this? 

Because it would work?  And it requires no language changes at all.

> I don't want to invent my own mechanisms, 

But you're asking *us* to invent a mechanism.

Right now the benefit of the invention we have is that it is entirely divorced from the language.  It is simply: you provide information *however* you want, and you generate *however* you want.

We only assume that the generated state of the world is valid and worth doing anything with.  We don't need any sort of language handshakes to invent.  It's all just an api.  Just like analyzers don't require anything special and can rev entirely separately from the lang, we'd like the same for generators.

**Comment by AaronRobinsonMSFT** [comment]
>In other words, the thing you're trying to fix/avoid here is precisely the use model we're trying heavily to move toward. the expectation is absolutely to not have the definitions or aything else informing the original source code. but all IDE features and whatnot only operate on the post-transformed code.

Then how is the generated code ever used? A key invariant of the source generators is that they can't modify any user defined code. How is this new code called/consumed without use of reflection?

**Comment by stephentoub** [comment]
> But you're asking us to invent a mechanism.

I'm asking the language to add a small feature that makes the source generator source way more usable and user-friendly, in my opinion.

> these scenarios have a suitable approach with the 'generate a new file' system. It may not be really glamorous, but i don't expect or want it to be. I'd rather have a simple system that can do this, instead of more and more complex features (like actually changing the language) in order to make this happen.

And the feedback I'm sharing, and that others here are sharing, and offline conversations from even more folks sharing this, is that this is very suboptimal.

**Comment by AaronRobinsonMSFT** [comment]
>We don't need any sort of language handshakes to invent. It's all just an api.

This isn't a requirement for sure. This simply makes the Source Generator feature usable in more scenarios (e.g. AOT).

**Comment by CyrusNajmabadi** [comment]
> I'm asking the language to add a small feature that makes the source generator way more usable and user-friendly, in my opinion.

Can you explain how it's more user friendly?   For users consuming the generator it's going to be the same right?  They'll simply see the final state of things and all things like intellisense, docs and whatnot will be the same.

In terms of people creating the generator, i'm happy to work on all sorts of IDE features to make that a pleasant experience.  but i'm not seeing the need for a language change.

**Comment by CyrusNajmabadi** [comment]
note: gotta run to the vet, will be back on later.

**Comment by CyrusNajmabadi** [comment]
> And the feedback I'm sharing, and that others here are sharing, and offline conversations from even more folks sharing this, is that this is very suboptimal.

BTW, can we bring the conversations online instead of offline, ideally with more examples.  I don't have the context for this, and that makes thinds difficult.  I can only drive off of what's been shown so far, and i'm not really seeing issues from either the consumption or productoin side.  The consumption side seems 100% ok.  The production side i can see possibly wanting to invest in.  But that investment could be IDE and not lang.

**Comment by AaronRobinsonMSFT** [comment]
>But that investment could be IDE ...

This does not align with a cross-platform scenario. .NET Core has cross-platform as a key focus and genesis for this request is at https://github.com/dotnet/runtime/pull/33742.

**Comment by CyrusNajmabadi** [comment]
I don't understand. Nothing about ides makes something not cross platform

**Comment by stephentoub** [comment]
> For users consuming the generator it's going to be the same right? They'll simply see the final state of things and all things like intellisense, docs and whatnot will be the same.

No.  In one case, the user writes:
```C#
/// <summary>Creates a regex that finds hello followed by world, with anything in the middle</summary>
[RegexGenerator("hello.*world", RegexOptions.IgnoreCase)]
internal static partial Regex HelloWorldRegex();
```
They write their helper method exactly like they'd write any other helper method, except the generator fills in the implementation for them.  There's very little here that's foreign, very few concepts to learn.  Super easy to teach.  They comment the method like they'd comment any other method.  They can control the visibility as they would with any other method.

In the other case, apparently they write:
```C#
// This is the documentation to generate on the member.  The source generator can look for this
// and move it over.  The source generator can totally make an informal mechanism to allow its domain
// to pass it the information it needs
[RegexGenerator("hello.*world", RegexOptions.IgnoreCase, nameof(HelloWorldRegex)]
```
learn that this is going to generate a method named HelloWorldRegex, learn that non-XML comments they randomly float above it are going to become comments on a generated method named HelloWorldRegex, have to look at the generated code (have to find the generated code) if they want to see the generated signature, ....  What if I wanted to make the method public... how do I convey that to the generator?  Another parameter and another ad-hoc convention?  What if I wanted the method to be instance rather than static?  With the partial method, I'm writing it out, I can just make it instance; how do I do that with the generator, pass in some reflection enum parameter to my attribute ctor?  And so on.

To me, the partial method non-void return support is a small feature that makes it possible to utilize the generator support in a much more understandable way.

**Comment by AaronRobinsonMSFT** [comment]
> I don't understand. Nothing about ides makes something not cross platform

Which IDE is going to support this? Is VS the first class IDE that we add support for? Do we instead focus on VS Code? In the meantime we have complicated Source Generator specific semantics to learn in order to actually become successful with the feature.

**Comment by CyrusNajmabadi** [comment]
> Which IDE is going to support this?

I would imagine all of them..  is there a reason they wouldn't?

**Comment by AaronRobinsonMSFT** [comment]
>is there a reason they wouldn't?

Because it isn't needed with a feature like this one. Both Source Generators and Code Analyzers can be run outside of an IDE without issue.

**Comment by CyrusNajmabadi** [comment]
>  There's very little here that's foreign, very few concepts to learn. Super easy to teach. 

I mean, wouldn't you just doc this with your attribute?

**Comment by agocke** [comment]
I'm not particularly convinced of the usability of the Regex use case, but I do suspect slightly augmenting partial methods would be useful.

My reasoning is: I could see many situations in which the generated member needs to have some sort of user-defined signature. Sure, every generator *could* invent their own type annotation system to describe the generated method signature... but we could also broaden the support for partial methods and effectively provide a first-class system for generation signatures.

**Comment by davidwrighton** [comment]
To pile on to the discussion, with @AaronRobinsonMSFT his feature depends not only on producing apis for use, but those apis will have complex types in them.

The idea is that it would be good to write
``` csharp
/// <summary>
///  Close a file handle
/// </summary>
[GeneratedDllImport("kernel32.dll")]
partial public void CloseHandle(SafeFileHandle fileHandle);
```

instead of something like
``` csharp
/// <summary>
///  Close a file handle
/// </summary>
[GeneratedDllImport("CloseHandle","kernel32.dll", new Type[]{typeof(void), typeof(SafeFileHandle)]
```

Which of these would be a more reasonable change for a developer used to developing P/Invoke apis today.

IMO the just drop an attribute into the world is reasonable when attaching something like a Xaml api to C#, but its less reasonable when the desire is specifically to define managed apis.

**Comment by stephentoub** [comment]
> Sure, every generator could invent their own type annotation system to describe the generated method signature... but we could also broaden the support for partial methods and effectively provide a first-class system for generation signatures.

Right, this is what I was getting at when talking about the visibility, static vs instance, XML docs, etc.

> I mean, wouldn't you just doc this with your attribute?

With more text that needs to be internalized and learned, sure.  With more unintuitive interactions in the IDE, sure. But I guess since we have documentation we don't need to make anything easier to use ever, since we can just explain away all complexity in documentation.

**Comment by CyrusNajmabadi** [comment]
Ok.  I hear you :-)

**Comment by CyrusNajmabadi** [comment]
Since SGs are all purely additive, it seems workable to have something indicate: this is a stub to be added to.

**Comment by agocke** [comment]
@CyrusNajmabadi Yes, although I particularly like that this wouldn't be a SG-specific feature. It feels like a pretty natural augmentation of existing partial method support.

**Comment by agocke** [comment]
For instance, it could be kind of cool to use them to represent a pseudo "signature" file

**Comment by CyrusNajmabadi** [comment]
So, feels like a better approach her might be too use partial (or a new keyword) for any type of symbol that will be generated. You can write the signatures of everything (types, members, etc), but it's an error to not have a generator supply the impl.

**Comment by CyrusNajmabadi** [comment]
I don't like this only being able partial methods.  People will want this for generated props for example :-)

**Comment by CyrusNajmabadi** [comment]
.d.cs files @agocke ? :-D

**Comment by agocke** [comment]
Definitely not `.csi`, nothing like F# signature files 😉

**Comment by jaredpar** [comment]
> This sounds like the original/replace approach that we decided to cancel.

The `original/replace` feature was quite a bit more invasive here. It allowed a generator to replace virtually any API call: `new`, methods, constructors, etc ... Further we paired it with making types implicitly `partial` to make generating `replace` methods much easier. This meant that generators could essentially override anything with no real opt-in from the user.

Note: it's been almost five years since we designed `original / replace` so I may be getting a few details wrong. Overall though I feel like that's the right picture (please correct me if I'm wrong)

The proposal here for `partial` methods is just relaxing two restrictions on an existing faeture. The feature itself though pretty much stays the same. The user declares the method and the generator fills it in pretty much the same as it's been doing since we invented `partial`. The real functional difference for the user is the method call is never erased.

To me these are very different features:
- `replace`: required no user opt-in and impossible to know at the call site that a method call was being replaced. The only way to know if it was replaced was to use "Go To Implementation" or do a lot of source code searching.
- `partial` method: explicit user opt-in and clear at the call site it's the `partial` method implementation that's being invoked. Yes still requires source code searching but at the same time the user knows there is something to search for.

I agree that there is a deal of overlap in the use cases, particularly for a few of the examples we've been talking about. But how the generator interacts with the user code and how it impacts their expectations are very different.

**Comment by CyrusNajmabadi** [comment]
@jaredpar That's fair.  I do want to reiterate my objection to this bit:

> The proposal here for partial methods is just relaxing two restrictions on an existing faeture.

I would not want the proposal to be just that.  Rather, if we're going to say "you can stub this out, and a generator will supply this" then we need to be able to do that for effectively anything a generator could generate (fortunately, that list is small and well defined i think).

i.e. all the argument @stephentoub, @davidwrighton and @AaronRobinsonMSFT made apply equally to properties as they do to methods.

So i would be totally ok with relaxing partial-on-methods, and enabling-partial-on-everything-else-with-a-body-or-initializer.

You basically can then write your expected signatures, have the generator supply them with real code this time, and have the compiler check this.

**Comment by CyrusNajmabadi** [comment]
open questions:

i view this as:  a way for code to signatures that have no other meaning except "this must be stubbed in".  The stubbed in code must match exactly (please, let's make sure param names match too).  

One area that i'm uncertain about is 'attributes'.  Would we ignore them on the 'stub' method?  it seems like that might make a bunch of sense in some cases.  i.e. if the attribute is just there to mark the stub for the generator to run on it.  We wouldn't want that attribute to be injected into code.

However, tehre are cases where you might want the attribute to stay.  For example:

```c#
[JsonName("foo"), SomeOtherGeneratorAttribute(...)]
public partian int Foo { get; }
```

here SomeOtherGeneratorAttribute would be to drive the generator, and would be stripped. but `[JsonName(...)]` would be something i would want to keep in.  There are several ways i could see this being done.

1. the attribute on the partial is ignored.  it's on the generator to copy the attributes over to the dest (excluding its own attributes).
2. attributes from partial+impl are merged (and maybe deduped?)
3. we keep all attributes, but we remove those deriving from some well known type (like `SourceGeneratorAttribute`).  

i think i'm leaning toward '1'.  but it does seem like something that would be easy for generators to screw up.  i could write a generator and totally not think about attributes.  and it would work for a bunch of consumers until someone tries to use it on a memb that has an existing attr.

**Comment by azany1** [comment]
In this case, I think '3’ is most “efficient”:  It puts the least burden (must derive any Generator-associated attribute from a single base) on the least number of actors (Generator authors).

-Damon

**Comment by jkotas** [comment]
The attribute from partial+impl are merged today. Why should source generators do anything else?

The C# compiler leaves a lot of other unnecessary stuff in the binary (e.g. private constants). If you want to get rid of the unnecessary stuff to make your binaries lean, you have to post-process it using ILLinker. Keeping all attributes around makes an existing problem a bit worse, it is not introducing a new problem.

**Comment by CyrusNajmabadi** [comment]
> Keeping all attributes around makes an existing problem a bit worse, it is not introducing a new problem.

Fair enough.  Seemed oogy to have an entirely source-oriented solution make it's way into metadata.  But if it doesn't offend people, it's not going to bother me that much :)

**Comment by YairHalberstadt** [comment]
> i.e. all the arguments made apply equally to properties as they do to methods.

To reiterate my point on https://github.com/dotnet/roslyn/pull/40162#issuecomment-562916696, one of the commonly requested use cases for source generators is generating calls to NotifyPropertyChanged from an auto property, without having to create a full property, via usage of an attribute.

This is tricky with the current proposal, but partial properties would make this trivial to do.

**Comment by YairHalberstadt** [comment]
> we keep all attributes, but we remove those deriving from some well known type (like SourceGeneratorAttribute).

You might not even need it to be a well known type. Conditional attribute should work:

```csharp
using System;
using System.Diagnostics;

[Conditional("sgadfbvsedbbs")]
public class SourceGeneratorAttribute : Attribute {

}

public class MyAttribute : SourceGeneratorAttribute 
{
     [MyAttribute]
     public void M() {}
}
```

The usage of MyAttribute is not emitted into metadata: https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxASwDYB8ACAmARgFgAoHABgAIcCA6AETQgHMA7Ae1jQGMoBuUqQDaAYXasAJmhhpxEDAAoARFGYQJAM2AA3KHAnBgUJQEoAuqRwBmankoBldgFcE3OAHE4rRBBjsEAIIwMAhowE7wlCCUQSFhEXCUAN6CJAC+qda2lACyAJ6xoeGR0Y4ubp7eCL7+hfGRpCkklC0tQvl1xXAWza3UNjgALLkKJskZ6UA==

(With thanks to @HaloFour)

**Comment by alrz** [comment]
Thinking out loud here, this could be actually used where you need replace/original, using some lightweight code convention that is completely in the control of the aurhor, take cached method example:

```cs
[Cached]
public partial int Compute(int i);
private int ComputeImpl(int i) {}
```
The generator will look for an `Impl` (or whatever) method and use it in the cached implementation.

That said, I support
partial properties idea, it would enable other (all?) scenarios previously discussed in the context of replace/original feature.

**Comment by RikkiGibson** [comment]
Right now the 'partial' modifier is required to be the last modifier before the type. As part of this feature should we relax this requirement?

**Comment by HaloFour** [comment]
@alrz 

That's an interesting use case.  I could see the (ab)use of partial properties as stubs for a generator to fill in a specific aspect:

```cs
[IPNC]
public partial class Foo {
    public partial string Name { get; set; }
}

// generated:
public partial class Foo : INotifyPropertyChanged {
    public event PropertyChangedEventHandler PropertyChanged; 

    private string _name;
    public partial string Name {
        get => _name;
        set {
            PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Name)));
        }
    }
}
```

The obvious problem with it is that it's not composable, you'd only ever be able to apply a single aspect.

I honestly hope that the team will reconsider `replace`/`original` or some other approach for AOP in the future.  IMO that's where the majority of compelling use cases are, and I think people are going to be quite frustrated once they realize the limitations of what will be shipped.  I personally feel that a compiler-provided approach via source generators would be significantly easier to use, reason about and debug than the existing AOP solutions.

**Comment by alrz** [comment]
@RikkiGibson 

>Right now the 'partial' modifier is required to be the last modifier before the type. As part of this feature should we relax this requirement?

I've done a standalone implementation of that long time ago and it's already being considered as a pre-requirement for records (https://github.com/dotnet/roslyn/pull/42097)

**Comment by alrz** [comment]
@HaloFour 

>The obvious problem with it is that it's not composable, you'd only ever be able to apply a single aspect.

Since the final method is being generated at last, you can build the composability into your generator with some code tricks, no?

**Comment by HaloFour** [comment]
@alrz 

> Since the final method is being generated at last, you can build the composability into your generator with some code tricks, no?

I could see that if you were composing multiple aspects supported by the same generator, but only if the generator were crafted to interpret those multiple aspects and emit a single implementation taking all of those aspects into consideration.  Other than that I don't see composability being practical.

> IIRC using multiple `replace` impls remained an open issue due to the ordering question.

Isn't it still an open question?  What happens if you reference multiple generators now and two of them want to emit the same types or members?  Can each generator see what the other generator has already emitted?  Do they run in a deterministic order and at deterministic times?

**Comment by alrz** [comment]
>Isn't it still an open question?

I don't think so. The issue seems to be resolved because we don't depend on it anymore.

> What happens if you reference multiple generators now and two of them want to emit the same types or members?

I wouldn't expect anything other than a compile-time error.

> Can each generator see what the other generator has already emitted? Do they run in a deterministic order and at deterministic times?

No, and the order is irrelevant as each generator can only see the original compilation from source.

**Comment by HaloFour** [comment]
@alrz 

> No, and the order is irrelevant as each generator can only see the original compilation from source.

Sounds like that would make it practically impossible to compose multiple aspects backed by different generators.

**Comment by alrz** [comment]
@HaloFour

>Sounds like that would make it practically impossible to compose multiple aspects backed by different generators.

One possible solution is to have a plugin-based framework on top of generators API to be able to chain multiple generator "plugins" with a defined execution order. Though I think that would be only widely usable if it's from MS.

**Comment by HaloFour** [comment]
@alrz 

Execution order and convention for emitting multiple partial methods into which each generator can put its implementation and chain to the next.  Doable, but I don't think it's practical without some major backer.  But maybe necessity will drive it.

**Comment by Neme12** [comment]
@alrz 
> Thinking out loud here, this could be actually used where you need replace/original, using some lightweight code convention that is completely in the control of the aurhor, take cached method example:
> 
> ```cs
> [Cached]
> public partial int Compute(int i);
> private int ComputeImpl(int i) {}
> ```
> 
> The generator will look for an `Impl` (or whatever) method and use it in the cached implementation.
> 
> That said, I support partial properties idea, it would enable other (all?) scenarios previously discussed in the context of replace/original feature.

I feel like that's not a good use case at all. If there is a convention, you could just declare `ComputeImpl` with the cached attribute on it and the source generator would create the `Compute` method that delegates to it. That way, you wouldn't have to declare the same method signature twice. This would just make you duplicate the signature without any benefit. The real use case for `partial` is where you don't already have any existing signature.
